src/btree/bt_delete.c

```diff
@@ -204,10 +204,26 @@ __wt_delete_page_rollback(WT_SESSION_IMPL *session, WT_REF *ref)
 			 */
 			__wt_free(session, ref->page_del->update_list);
 			__wt_free(session, ref->page_del);
-			goto done;
+			return;
 		}
+		/*
+		 * We wait for the change in page state, yield before retrying,
+		 * and if we've yielded enough times, start sleeping so we don't
+		 * burn CPU to no purpose.
+		 */
+		if (yield_count < WT_THOUSAND) {
+			yield_count++;
+			__wt_yield();
+			continue;
+		}
+
+		sleep_count = WT_MIN(sleep_count + WT_THOUSAND,
+		    10 * WT_THOUSAND);
+		WT_STAT_CONN_INCRV(session, page_del_rollback_blocked,
+		    sleep_count);
+		__wt_sleep(0, sleep_count);
```

Comment:
This logic is repeated in three places now `__wt_page_in_func` and the two here, it'd be good if they all shared the same implementation. Maybe call it `inline void __wt_ref_state_wait_backoff(WT_SESSION_IMPL *session, uin32_t *yield_count, uint32_t *sleep_count, int64_t stat_to_increment)`

You may need yet more parameters (`bool should_evict`) to support the structure in `__wt_page_in_func`, but I think you might also be able to fold the `yield_count` and `sleep_count` into a single parameter.